pkg/lib/cockpit-components-context-menu.jsx

```diff
@@ -0,0 +1,98 @@
+/*
+ * This file is part of Cockpit.
+ *
+ * Copyright (C) 2019 Red Hat, Inc.
+ *
+ * Cockpit is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation; either version 2.1 of the License, or
+ * (at your option) any later version.
+ *
+ * Cockpit is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with Cockpit; If not, see <http://www.gnu.org/licenses/>.
+ */
+
+import React from "react";
+
+import "context-menu.css";
+
+export class ContextMenu extends React.Component {
+    constructor() {
+        super();
+        this.state = { visible: false };
+        this._handleContextMenu = this._handleContextMenu.bind(this);
+        this._handleClick = this._handleClick.bind(this);
+    }
+
+    componentDidMount() {
+        document.addEventListener('contextmenu', this._handleContextMenu);
+        document.addEventListener('click', this._handleClick);
+    }
+
+    componentWillUnmount() {
+        document.removeEventListener('contextmenu', this._handleContextMenu);
+        document.removeEventListener('click', this._handleClick);
+    }
+
+    _handleContextMenu(event) {
+        event.preventDefault();
+
+        this.setState({ visible: true });
+
+        const clickX = event.clientX;
+        const clickY = event.clientY;
+        const screenW = window.innerWidth;
+        const screenH = window.innerHeight;
+        const rootW = this.root.offsetWidth;
+        const rootH = this.root.offsetHeight;
+
+        const right = (screenW - clickX) > rootW;
+        const left = !right;
+        const top = (screenH - clickY) > rootH;
+        const bottom = !top;
+
+        if (right) {
+            this.root.style.left = `${clickX + 5}px`;
+        }
+
+        if (left) {
+            this.root.style.left = `${clickX - rootW - 5}px`;
+        }
+
+        if (top) {
+            this.root.style.top = `${clickY + 5}px`;
+        }
+
+        if (bottom) {
+            this.root.style.top = `${clickY - rootH - 5}px`;
```

Comment:
This looks like CSS attributes? Doesn't that trigger a Content Security Policy error?

Reply:
I have not noticed that CSP would happen. Most likely because it is only set when component is created.

Reply:
It's less a question of when it happens, but where -- Neither HTML or JS is usually allowed to inject CSS. But if it works, fine (I don't claim to understand all details of CSP).